src/allmydata/test/test_openmetrics.py

```diff
@@ -0,0 +1,118 @@
+           'storage_server.latencies.get.90_0_percentile': 0.0002999305725097656,
+           'storage_server.latencies.get.01_0_percentile': 0.0001239776611328125,
+           'cpu_monitor.total': 641.4941180000001,
+           'storage_server.latencies.write.samplesize': 1000,
+           'storage_server.latencies.write.95_0_percentile': 9.489059448242188e-05,
+           'storage_server.latencies.read.50_0_percentile': 6.890296936035156e-05,
+           'storage_server.latencies.writev.01_0_percentile': 0.00033211708068847656,
+           'storage_server.latencies.read.10_0_percentile': 3.0994415283203125e-05,
+           'storage_server.latencies.allocate.10_0_percentile': 0.0004949569702148438,
+           'storage_server.reserved_space': 0,
+           'storage_server.disk_avail': 103289454592,
+           'storage_server.latencies.write.99_0_percentile': 0.00011301040649414062,
+           'storage_server.latencies.write.90_0_percentile': 9.083747863769531e-05,
+           'cpu_monitor.5min_avg': 0.0002370666691157502,
+           'storage_server.latencies.write.mean': 5.8008909225463864e-05,
+           'storage_server.latencies.readv.50_0_percentile': 0.00033020973205566406,
+           'storage_server.latencies.close.99_0_percentile': 0.0004038810729980469,
+           'storage_server.allocated': 0,
+           'storage_server.latencies.writev.99_0_percentile': 0.0007710456848144531,
+           'storage_server.latencies.readv.99_9_percentile': 0.0004780292510986328,
+           'storage_server.latencies.read.samplesize': 170,
+           'storage_server.latencies.allocate.samplesize': 406,
+           'storage_server.latencies.allocate.95_0_percentile': 0.0008411407470703125
+         }, 'counters': {
+           'storage_server.writev': 309,
+           'storage_server.bytes_added': 197836146,
+           'storage_server.close': 326,
+           'storage_server.readv': 14299,
+           'storage_server.allocate': 406,
+           'storage_server.read': 170,
+           'storage_server.write': 3775,
+           'storage_server.get': 472}
+         }
+        return stats
+
+class OpenMetrics(unittest.TestCase):
+    def test_spec_compliance(self):
+        """
+        Does our output adhere to the OpenMetrics spec?
+        https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md
+        """
+        req = mock.Mock()
+        stats = mock.Mock()
+        stats._provider = FakeStatsProvider()
+        metrics = Statistics.render_OPENMETRICS(stats, req)
+
+        # "The content type MUST be..."
+        req.setHeader.assert_called_with(
+            "content-type", "application/openmetrics-text; version=1.0.0; charset=utf-8"
+        )
+
+        # The parser throws if it does not like its input.
+        # Wrapped in a list() to drain the generator.
+        families = list(parser.text_string_to_metric_families(metrics))
```

Comment:
Can we assert that the resulting value somehow overlaps with the inputs supplied?

Reply:
Hm.
The parser should be testing our output here, not the other way around.
I added an ad hoc check.